include/ghex/communication_object_ipr.hpp

```diff
@@ -0,0 +1,564 @@
+             * @tparam DomainIdType domain id type*/
+            template<typename Communicator, typename GridType, typename DomainIdType>
+            class communication_object_ipr {
+
+                private:
+
+                    // friend class // TO DO: is friend needed?
+                    friend class communication_handle_ipr<Communicator,GridType,DomainIdType>;
+
+                public:
+
+                    // member types
+
+                    /** @brief handle type returned by exhange operation */
+                    using communicator_type = Communicator;
+                    using grid_type = GridType;
+                    using domain_id_type = DomainIdType;
+                    using handle_type = communication_handle_ipr<communicator_type,grid_type,domain_id_type>;
+                    using pattern_type = pattern<communicator_type,grid_type,domain_id_type>;
+                    using pattern_container_type = pattern_container<communicator_type,grid_type,domain_id_type>;
+
+                    template<typename D, typename F>
+                    using buffer_info_type = buffer_info<pattern_type,D,F>;
+
+                private:
+
+                    // member types
+
+                    using address_type = typename communicator_type::address_type;
+                    using index_container_type = typename pattern_type::index_container_type;
+                    using pack_function_type = std::function<void(void*, const index_container_type&, void*)>; // last argument intended as CUDA stream
+
+                    /** @brief pair of domain ids + tag id with ordering */
+                    struct domain_id_pair_and_tag {
+
+                            domain_id_type first_id;
+                            domain_id_type second_id;
+                            int tag;
+
+                            bool operator<(const domain_id_pair_and_tag& other) const noexcept {
+                                return (first_id < other.first_id ? true :
+                                    (first_id > other.first_id ? false :
+                                        (second_id < other.second_id ? true :
+                                            (second_id > other.second_id ? false :
+                                                (tag < other.tag)))));
+                            }
+
+                    };
+
+                    /** @brief Holds a pointer to a set of iteration spaces and a callback function pointer
+                     * which is used to store a field's pack or unpack member function.
+                     * This class also stores the offset in the serialized buffer in bytes.
+                     * The type-erased field_ptr member is only used for the gpu-vector-interface. // TO DO: verify is still needed in this case
+                     * @tparam Function either pack or unpack function pointer type */
+                    template<typename Function>
+                    struct field_info {
```

Comment:
Also this struct is the same as the co_2. Would it make sense to use some composition to reuse some code? (how much code would be saved by doing so?)

Reply:
Moved to refactoring issue.